## Energy Consumption Forecasting for Indian States


#### Life cycle of this project

- Understanding the Problem Statement
- Data Collection
- Data Checks to perform
- Exploratory data analysis
- Data Pre-Processing
- Model Training
- Choose best model

### 1) Problem statement
- India’s diverse energy consumption patterns across states make it difficult for grid operators to efficiently balance supply and demand. This project aims to develop a forecasting model using historical state-wise energy consumption data to improve resource allocation, prevent blackouts, and promote sustainable energy practices.



### 2) Data Collection
- Dataset Source - https://www.kaggle.com/datasets/twinkle0705/state-wise-power-consumption-in-india
- This data is in the form of a time series for a period of 17 months beginning from 2nd Jan 2019 till 23rd May 2020.
- It has 16.6k rows of data.

### 2.1 Import Data and Required Packages
####  Importing Pandas, Numpy, Matplotlib, Seaborn and Warings Library.

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

#### Import the CSV Data as Pandas DataFrame

In [3]:
df = pd.read_csv('../data/long_data.csv')

#### Show Top 5 Records

In [4]:
df.head()

States Regions   latitude  longitude                Dates  Usage
0     Punjab      NR  31.519974  75.980003  02/01/2019 00:00:00  119.9
1    Haryana      NR  28.450006  77.019991  02/01/2019 00:00:00  130.3
2  Rajasthan      NR  26.449999  74.639981  02/01/2019 00:00:00  234.1
3      Delhi      NR  28.669993  77.230004  02/01/2019 00:00:00   85.8
4         UP      NR  27.599981  78.050006  02/01/2019 00:00:00  313.9

#### Shape of the dataset

In [5]:
df.columns

Index(['States', 'Regions', 'latitude', 'longitude', 'Dates', 'Usage'], dtype='object')

### 2.2 Dataset information

- States : Indian States (33 unique values including union territories)
- Regions : Indian Region from where the data is  -> (NR, NER, WR, SR, ER)
- latitude : Latitude of the location
- longitude : Longitude of the location 
- Dates : Comple date and time (DD/MM/YYYY hh:mm:ss)
- Usage : Electricity usage in Mega Units (MU)

### 3. Data Checks to perform

- Check Missing values
- Check Duplicates
- Check data type
- Check the number of unique values of each column
- Check statistics of data set
- Check various categories present in the different categorical column

### 3.1 Check Missing values

In [6]:
df.isna().sum()

States       0
Regions      0
latitude     0
longitude    0
Dates        0
Usage        0
dtype: int64

#### There are no missing values in the data set

### 3.2 Check Duplicates

In [9]:
df.duplicated().sum()

np.int64(12)

#### There are 12 duplicates values in the data set

In [11]:
duplicates = df[df.duplicated(keep=False)]
duplicates

States Regions   latitude  longitude                Dates  \
6032  Arunachal Pradesh     NER  27.100399  93.616601  08/07/2019 00:00:00   
6065  Arunachal Pradesh     NER  27.100399  93.616601  08/07/2019 00:00:00   
6102            Mizoram     NER  23.710399  92.720015  09/07/2019 00:00:00   
6135            Mizoram     NER  23.710399  92.720015  09/07/2019 00:00:00   
6163             Sikkim      ER  27.333330  88.616647  10/07/2019 00:00:00   
6167          Meghalaya     NER  25.570492  91.880014  10/07/2019 00:00:00   
6168            Mizoram     NER  23.710399  92.720015  10/07/2019 00:00:00   
6169           Nagaland     NER  25.666998  94.116570  10/07/2019 00:00:00   
6170            Tripura     NER  23.835404  91.279999  10/07/2019 00:00:00   
6196             Sikkim      ER  27.333330  88.616647  10/07/2019 00:00:00   
6200          Meghalaya     NER  25.570492  91.880014  10/07/2019 00:00:00   
6201            Mizoram     NER  23.710399  92.720015  10/07/2019 00:00:00   
6202           Nagaland     NER  25.666998  94.116570  10/07/2019 00:00:00   
6203            Tripura     NER  23.835404  91.279999  10/07/2019 00:00:00   
6275        Uttarakhand      NR  30.320409  78.050006  12/07/2019 00:00:00   
6290              Pondy      SR  11.934994  79.830000  12/07/2019 00:00:00   
6296  Arunachal Pradesh     NER  27.100399  93.616601  12/07/2019 00:00:00   
6301           Nagaland     NER  25.666998  94.116570  12/07/2019 00:00:00   
6302            Tripura     NER  23.835404  91.279999  12/07/2019 00:00:00   
6308        Uttarakhand      NR  30.320409  78.050006  12/07/2019 00:00:00   
6323              Pondy      SR  11.934994  79.830000  12/07/2019 00:00:00   
6329  Arunachal Pradesh     NER  27.100399  93.616601  12/07/2019 00:00:00   
6334           Nagaland     NER  25.666998  94.116570  12/07/2019 00:00:00   
6335            Tripura     NER  23.835404  91.279999  12/07/2019 00:00:00   

      Usage  
6032    1.4  
6065    1.4  
6102    1.4  
6135    1.4  
6163    1.5  
6167    4.1  
6168    1.4  
6169    1.8  
6170    2.9  
6196    1.5  
6200    4.1  
6201    1.4  
6202    1.8  
6203    2.9  
6275   34.1  
6290    7.4  
6296    2.1  
6301    2.1  
6302    3.5  
6308   34.1  
6323    7.4  
6329    2.1  
6334    2.1  
6335    3.5

In [14]:
df_cleaned = df.drop_duplicates()
df_cleaned.duplicated().sum()
df = df_cleaned

### 3.3 Check data types

In [15]:
# Check Null and Dtypes
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16587 entries, 0 to 16598
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   States     16587 non-null  object 
 1   Regions    16587 non-null  object 
 2   latitude   16587 non-null  float64
 3   longitude  16587 non-null  float64
 4   Dates      16587 non-null  object 
 5   Usage      16587 non-null  float64
dtypes: float64(3), object(3)
memory usage: 907.1+ KB


### 3.4 Checking the number of unique values of each column

In [16]:
df.nunique()

States         33
Regions         5
latitude       33
longitude      32
Dates         498
Usage        3627
dtype: int64

### 3.5 Check statistics of data set

In [17]:
df.describe()

latitude     longitude         Usage
count  16587.000000  16587.000000  16587.000000
mean      23.177156     81.788486    103.072539
std        6.147597      7.256201    116.056017
min        8.900373     71.192400      0.300000
25%       19.820430     76.569993      6.700000
50%       23.835404     78.570026     64.600000
75%       27.333330     88.329947    174.000000
max       33.450000     94.216667    522.100000

#### Insight

- Usage shows high variability, with a standard deviation (116.056017) larger than its mean (103.072539)
- While the minimum usage is very low (0.300000), the maximum (522.100000) is significantly higher than the 75th percentile (174.000000)
- The median usage (64.600000) is much lower than the mean, suggesting a right-skewed distribution
- Latitude has the smallest range (24.549627°) compared to longitude (23.024267°), indicating a more compact north-south spread.

### 3.7 Exploring Data

In [18]:
df.head()

States Regions   latitude  longitude                Dates  Usage
0     Punjab      NR  31.519974  75.980003  02/01/2019 00:00:00  119.9
1    Haryana      NR  28.450006  77.019991  02/01/2019 00:00:00  130.3
2  Rajasthan      NR  26.449999  74.639981  02/01/2019 00:00:00  234.1
3      Delhi      NR  28.669993  77.230004  02/01/2019 00:00:00   85.8
4         UP      NR  27.599981  78.050006  02/01/2019 00:00:00  313.9

In [19]:
print("Categories in 'States' variable:  ",end=" " )
print(df['States'].unique())

print("Categories in 'Regions' variable:  ",end=" ")
print(df['Regions'].unique())

print("latitude: ",end=" " )
print(df['latitude'].unique())

print("longitude: ",end=" " )
print(df['longitude'].unique())

print("Dates:  ",end=" " )
print(df['Dates'].unique())

print("Usage:  ",end=" " )
print(df['Usage'].unique())

Categories in 'States' variable:   ['Punjab' 'Haryana' 'Rajasthan' 'Delhi' 'UP' 'Uttarakhand' 'HP' 'J&K'
 'Chandigarh' 'Chhattisgarh' 'Gujarat' 'MP' 'Maharashtra' 'Goa' 'DNH'
 'Andhra Pradesh' 'Telangana' 'Karnataka' 'Kerala' 'Tamil Nadu' 'Pondy'
 'Bihar' 'Jharkhand' 'Odisha' 'West Bengal' 'Sikkim' 'Arunachal Pradesh'
 'Assam' 'Manipur' 'Meghalaya' 'Mizoram' 'Nagaland' 'Tripura']
Categories in 'Regions' variable:   ['NR' 'WR' 'SR' 'ER' 'NER']
latitude:  [31.51997398 28.45000633 26.44999921 28.6699929  27.59998069 30.32040895
 31.10002545 33.45       30.71999697 22.09042035 22.2587     21.30039105
 19.25023195 15.491997   20.26657819 14.7504291  18.1124     12.57038129
  8.90037274 12.92038576 11.93499371 25.78541445 23.80039349 19.82042971
 22.58039044 27.3333303  27.10039878 26.7499809  24.79997072 25.57049217
 23.71039899 25.6669979  23.83540428]
longitude:  [75.98000281 77.01999101 74.63998124 77.23000403 78.05000565 77.16659704
 76.24       76.78000565 82.15998734 71.1924     76.13

In [20]:
# define numerical & categorical columns
numeric_features = [feature for feature in df.columns if df[feature].dtype != 'O']
categorical_features = [feature for feature in df.columns if df[feature].dtype == 'O']

# print columns
print('We have {} numerical features : {}'.format(len(numeric_features), numeric_features))
print('\nWe have {} categorical features : {}'.format(len(categorical_features), categorical_features))

We have 3 numerical features : ['latitude', 'longitude', 'Usage']

We have 3 categorical features : ['States', 'Regions', 'Dates']


In [21]:
df.head(2)

States Regions   latitude  longitude                Dates  Usage
0   Punjab      NR  31.519974  75.980003  02/01/2019 00:00:00  119.9
1  Haryana      NR  28.450006  77.019991  02/01/2019 00:00:00  130.3

### 3.8 Separating Date and Time

In [25]:
# Convert 'Dates' to datetime using the specific format
df['Dates'] = pd.to_datetime(df['Dates'], format='%d/%m/%Y %H:%M:%S')

# Separate date and time
df['Date'] = pd.to_datetime(df['Dates']).dt.date
df['Time'] = pd.to_datetime(df['Dates']).dt.time

# Drop the original 'Dates' column
df = df.drop('Dates', axis=1)

# Reorder columns
column_order = ['States', 'Regions', 'latitude', 'longitude', 'Date', 'Time', 'Usage']
df = df[column_order]

df.tail()

States Regions   latitude  longitude        Date      Time  Usage
16594    Manipur     NER  24.799971  93.950017  2020-12-05  00:00:00    2.5
16595  Meghalaya     NER  25.570492  91.880014  2020-12-05  00:00:00    5.8
16596    Mizoram     NER  23.710399  92.720015  2020-12-05  00:00:00    1.6
16597   Nagaland     NER  25.666998  94.116570  2020-12-05  00:00:00    2.1
16598    Tripura     NER  23.835404  91.279999  2020-12-05  00:00:00    3.3

In [26]:
usage_less_20 = df[df['Usage'] <= 20].count()


print(f'Electricity usage less than 20 MU: {usage_less_20}')

Electricity usage less than 20 MU: States       5659
Regions      5659
latitude     5659
longitude    5659
Date         5659
Time         5659
Usage        5659
dtype: int64
